### ISSUE TITLE
Generic argument inference for a type, based on initializer call

### ISSUE THREAD

**Issue by tangent-vector** [shader-slang/slang#327]
This is not something we need to get to right away, but if I have a generic type, with user-written initializers:

```hlsl
struct Foo<T>
{
    init(T val) { ... }
}
```

there may be cases where the correct type argument can be inferred at a use site:

```hlsl
int x = 123;
doSomething( Foo(x) ); // user clearly wants `Foo<int>`
```

This case trips up our current approach to generic argument inference, since we end up wanting to create a specialized decl-ref to `Foo` before we can look at its initializer(s), but we can't actually perform the inference step until we are looking at the candidate initializer(s).

This would have a low priority to fix, so I'm just documenting this gap for the long term.
